pkg/genyaml/genyaml.go

```diff
@@ -0,0 +1,404 @@
+	cm := &CommentMap{
+		comments: make(map[string]map[string]Comment),
+	}
+
+	for _, path := range paths {
+		cm.AddPath(path)
+	}
+
+	return cm
+}
+
+// Comment is an abstract structure for storing parsed AST comments decorated with contextual information.
+type Comment struct {
+	// Type is the underlying type of the identifier associated with the comment.
+	Type string
+	// IsObj determines if the underlying type is a object type (e.g. struct) or primitive type (e.g. string).
+	IsObj bool
+	// Doc is a comment string parsed from the AST of a node.
+	Doc string
+}
+
+// marshal marshals the object into JSON then converts JSON to YAML and returns the YAML.
+func marshal(o interface{}) ([]byte, error) {
+	j, err := json.Marshal(o)
+	if err != nil {
+		return nil, fmt.Errorf("error marshaling into JSON: %v", err)
+	}
+
+	y, err := jsonToYaml(j)
+	if err != nil {
+		return nil, fmt.Errorf("error converting JSON to YAML: %v", err)
+	}
+
+	return y, nil
+}
+
+// jsonToYaml Converts JSON to YAML.
+func jsonToYaml(j []byte) ([]byte, error) {
+	// Convert the JSON to an object.
+	var jsonObj interface{}
+	// We are using yaml.Unmarshal here (instead of json.Unmarshal) because the
+	// Go JSON library doesn't try to pick the right number type (int, float,
+	// etc.) when unmarshalling to interface{}, it just picks float64
+	// universally. go-yaml does go through the effort of picking the right
+	// number type, so we can preserve number type throughout this process.
+	err := yaml3.Unmarshal(j, &jsonObj)
+	if err != nil {
+		return nil, err
+	}
+
+	// marshal this object into YAML.
+	return yaml3.Marshal(jsonObj)
+}
+
+// astFrom takes a path to a Go file and returns the abstract syntax tree (AST) for that file.
+func astFrom(path string) (*doc.Package, error) {
+	fset := token.NewFileSet()
+	m := make(map[string]*ast.File)
+
+	f, err := parser.ParseFile(fset, path, nil, parser.ParseComments)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse file to AST from path: %s", path)
+	}
+
+	m[path] = f
+	apkg, _ := ast.NewPackage(fset, m, nil, nil)
+
+	astDoc := doc.New(apkg, "", 0)
+	if astDoc == nil {
+		return nil, fmt.Errorf("unable to parse AST documentation from path: %s", path)
+	}
+
+	return astDoc, nil
+}
+
+// fmtRawDoc formats/sanitizes a Go doc string removing TODOs, newlines, whitespace, and various other characters from the resultant string.
+func fmtRawDoc(rawDoc string) string {
+	var buffer bytes.Buffer
+	delPrevChar := func() {
+		if buffer.Len() > 0 {
+			buffer.Truncate(buffer.Len() - 1) // Delete the last " " or "\n".
+		}
+	}
+
+	// Ignore all lines after ---.
+	rawDoc = strings.Split(rawDoc, "---")[0]
+
+	for _, line := range strings.Split(rawDoc, "\n") {
+		line = strings.TrimRight(line, " ")
+		leading := strings.TrimLeft(line, " ")
+		switch {
+		case len(line) == 0: // Keep paragraphs
+			delPrevChar()
+			buffer.WriteString("\n\n")
+		case strings.HasPrefix(leading, "TODO"): // Ignore one line TODOs.
```

Comment:
This is behavior taken directly from [kubernetes/kubernetes](github.com/kubernetes/kubernetes) swagger doc generation [script](https://github.com/kubernetes/kubernetes/blob/master/hack/lib/swagger.sh#L44-L46). I have no problem changing if this does not make sense for out use-case.

If we do allow multiline TODOs, it could get difficult from a parsing perspective, to determine when it ends unless we use a sentinel or restrict to block comments. I can investigate the various options.

Reply:
I think the problem is that people aren't going to know that multiline TODOs are forbidden and so someone will inevitably make one.

That said, this is probably a Good Enough heuristic.

Reply:
I will investigate block comment TODOs. At the very least, I will disambiguate this in the README to make the usage clear.

Reply:
I added detailed usage to the [README](https://github.com/kubernetes/test-infra/pull/13637/commits/5b18160d003f4e0954ca7ebddbc10240f62f9633#diff-809bdf35d841e272801a6cc6b6737dc8R8-R111) outlining exactly what happens to comments parsed by this utility.